### PR TITLE
Add periodic aggregation of datapath notifications

### PR DESCRIPTION
**Pull request by joestringer**
Requires ~~#4867~~, ~~#4874~~ to reduce complexity enough to support this new functionality on Linux 4.9.17.

This is the next step after ~~#4828~~. It adds a new `MonitorAggregationLevel`, `Medium`. This will reduce notifications further, such that notifications will only be sent if there is new information (such as TCP flags that were not previously seen), or a packet is received and a notification has not been sent in the past five seconds. Future work may make this period more configurable.

---

Example monitor output for a standard TCP curl for a basic service (from the docker guide on docs.cilium.io) with both endpoints in Medium aggregation mode:

```
-> endpoint 28601 flow 0x41d01aa identity 54742->54742 state new ifindex lxc33fa9: 10.11.66.250:40210 -> 10.11.176.4:80 tcp SYN       
-> endpoint 44008 flow 0xbd455b43 identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40210 tcp SYN, ACK
-> endpoint 28601 flow 0x41d01aa identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40210 -> 10.11.176.4:80 tcp ACK
-> endpoint 28601 flow 0x41d01aa identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40210 -> 10.11.176.4:80 tcp ACK
-> endpoint 44008 flow 0xbd455b43 identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40210 tcp ACK
-> endpoint 28601 flow 0x41d01aa identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40210 -> 10.11.176.4:80 tcp ACK, FIN
-> endpoint 44008 flow 0xbd455b43 identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40210 tcp ACK, FIN
```

Compare with aggregation mode none:

```
<- endpoint 44008 flow 0xac35ebce identity 54742->0 state new ifindex 0: 10.11.66.250:40710 -> 10.11.176.4:80 tcp SYN
-> endpoint 28601 flow 0xac35ebce identity 54742->54742 state new ifindex lxc33fa9: 10.11.66.250:40710 -> 10.11.176.4:80 tcp SYN
<- endpoint 28601 flow 0xef0d08df identity 54742->0 state new ifindex 0: 10.11.176.4:80 -> 10.11.66.250:40710 tcp SYN, ACK
-> endpoint 44008 flow 0xef0d08df identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40710 tcp SYN, ACK
<- endpoint 44008 flow 0xac35ebce identity 54742->0 state new ifindex 0: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
-> endpoint 28601 flow 0xac35ebce identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
<- endpoint 44008 flow 0xac35ebce identity 54742->0 state new ifindex 0: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
-> endpoint 28601 flow 0xac35ebce identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
<- endpoint 28601 flow 0xef0d08df identity 54742->0 state new ifindex 0: 10.11.176.4:80 -> 10.11.66.250:40710 tcp ACK
-> endpoint 44008 flow 0xef0d08df identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40710 tcp ACK
<- endpoint 28601 flow 0xef0d08df identity 54742->0 state new ifindex 0: 10.11.176.4:80 -> 10.11.66.250:40710 tcp ACK
-> endpoint 44008 flow 0xef0d08df identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40710 tcp ACK
<- endpoint 44008 flow 0xac35ebce identity 54742->0 state new ifindex 0: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
-> endpoint 28601 flow 0xac35ebce identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
<- endpoint 44008 flow 0xac35ebce identity 54742->0 state new ifindex 0: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK, FIN
-> endpoint 28601 flow 0xac35ebce identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK, FIN
<- endpoint 28601 flow 0xef0d08df identity 54742->0 state new ifindex 0: 10.11.176.4:80 -> 10.11.66.250:40710 tcp ACK, FIN
-> endpoint 44008 flow 0xef0d08df identity 54742->54742 state reply ifindex lxc8736a: 10.11.176.4:80 -> 10.11.66.250:40710 tcp ACK, FIN
<- endpoint 44008 flow 0xac35ebce identity 54742->0 state new ifindex 0: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
-> endpoint 28601 flow 0xac35ebce identity 54742->54742 state established ifindex lxc33fa9: 10.11.66.250:40710 -> 10.11.176.4:80 tcp ACK
```

Upgrade impact is that the conntrack map will be recreated on startup, which may cause temporary disruption of ongoing connections.

Related: #4080 
Supersedes ~~#4875~~.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4875)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Rebased against #4917, retrying CI.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Rebased for latest CI fixes.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
I opened a fresh PR with just the BPF-side changes, which I suspect will pass CI (modulo flakes), then we can look at the userspace side independently to figure out what's going on with these failures.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit #4946:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5270

**Comment by joestringer** [comment]
Other failure:
```
Stacktrace
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
Failed to produce from empire-hq on topic deathstar-plans
Expected
    <*errors.errorString | 0xc42147a210>: {
        s: "ExecKafkaPodCmd: command 'kubectl exec -n default empire-hq-2917507631-wpc69 sh -- -c \"echo 'deathstar reactor design v3' | ./kafka-produce.sh --topic deathstar-plans\"' failed '>>' || '[2018-07-25 05:35:57,910] ERROR Error when sending message to topic deathstar-plans with key: null, value: 27 bytes with error: (org.apache.kafka.clients.producer.internals.ErrorLoggingCallback)\norg.apache.kafka.common.errors.TimeoutException: Expiring 1 record(s) for deathstar-plans-0: 1580 ms has passed since batch creation plus linger time\n'",
    }
to be nil
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/KafkaPolicies.go:230
Standard Output
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 1
Number of "Cilium API handler panicked" in logs: 0
Cilium pods: [cilium-d7wbh cilium-hvlnt]
Netpols loaded: 
CiliumNetworkPolicies loaded: default::cnp-second-namespace default::kafka-sw-security-policy 
Endpoint Policy Enforcement:
     empire-backup-2986355077-bg5gq         =>   none
     empire-outpost-9999-4143713519-jpc0n   =>   none
     kafka-broker-2873197211-k2nls          =>   ingress
     cilium-health-k8s2                     =>   none
     prometheus-core-3506541099-4n1kl       =>   none
     empire-hq-2917507631-wpc69             =>   none
     empire-outpost-8888-3627733903-b33z2   =>   none
     zookeeper-1032194181-jv7d4             =>   none
     cilium-health-k8s1                     =>   none
     kube-dns-2552250787-lztr7              =>   none
     microscope                             =>   none
Cilium agent "cilium-d7wbh": Status: Ok  Health: Ok Nodes "k8s1 k8s2" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 47 Failed 0
Cilium agent "cilium-hvlnt": Status: Ok  Health: Ok Nodes "k8s2 k8s1" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 39 Failed 0

Standard Error
STEP: Wait for Kafka broker to be up
STEP: Creating new kafka topic empire-announce
STEP: Creating new kafka topic deathstar-plans
STEP: Waiting for DNS to resolve within pods for kafka-service
STEP: Testing basic Kafka Produce and Consume
STEP: Apply L7 kafka policy and wait
STEP: validate that the pods have the correct policy
STEP: Testing Kafka L7 policy enforcement status
=== Test Finished at 2018-07-25T05:36:04Z====
===================== TEST FAILED =====================
cmd: kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                                   READY     STATUS        RESTARTS   AGE       IP              NODE
default       empire-backup-2986355077-bg5gq         1/1       Running       0          2m        10.10.0.66      k8s1
default       empire-hq-2917507631-wpc69             1/1       Running       0          2m        10.10.0.147     k8s1
default       empire-outpost-8888-3627733903-b33z2   1/1       Running       0          2m        10.10.1.90      k8s2
default       empire-outpost-9999-4143713519-jpc0n   1/1       Running       0          2m        10.10.0.35      k8s1
default       kafka-broker-2873197211-k2nls          1/1       Running       0          2m        10.10.0.199     k8s1
default       zookeeper-1032194181-jv7d4             1/1       Running       0          2m        10.10.1.69      k8s2
kube-system   cilium-d7wbh                           1/1       Running       0          13m       192.168.36.11   k8s1
kube-system   cilium-hvlnt                           1/1       Running       0          13m       192.168.36.12   k8s2
kube-system   etcd-k8s1                              1/1       Running       0          49m       192.168.36.11   k8s1
kube-system   kube-apiserver-k8s1                    1/1       Running       0          49m       192.168.36.11   k8s1
kube-system   kube-controller-manager-k8s1           1/1       Running       1          49m       192.168.36.11   k8s1
kube-system   kube-dns-2552250787-lztr7              3/3       Running       0          40m       10.10.1.32      k8s2
kube-system   kube-proxy-7p0l4                       1/1       Running       0          42m       192.168.36.12   k8s2
kube-system   kube-proxy-zx0sc                       1/1       Running       0          50m       192.168.36.11   k8s1
kube-system   kube-scheduler-k8s1                    1/1       Running       0          49m       192.168.36.11   k8s1
kube-system   microscope                             0/1       Terminating   0          1m        <none>          k8s1
prometheus    prometheus-core-3506541099-4n1kl       1/1       Running       0          42m       10.10.1.102     k8s2

cmd: kubectl exec -n kube-system cilium-d7wbh -- cilium service list
ID   Frontend            Backend                   
1    10.96.44.22:9090    1 => 10.10.1.102:9090     
2    10.96.0.1:443       1 => 192.168.36.11:6443   
3    10.96.0.10:53       1 => 10.10.1.32:53        
9    10.97.166.56:2181   1 => 10.10.1.69:2181      

cmd: kubectl exec -n kube-system cilium-d7wbh -- cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                       IPv6                 IPv4          STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                                       
3760       Disabled           Disabled          11228      k8s:app=empire-hq                                 f00d::a0a:0:0:eb0    10.10.0.147   ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                              
                                                           k8s:io.kubernetes.pod.namespace=default                                                      
                                                           k8s:zgroup=kafkaTestApp                                                                      
6833       Disabled           Disabled          60296      k8s:app=empire-backup                             f00d::a0a:0:0:1ab1   10.10.0.66    ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                              
                                                           k8s:io.kubernetes.pod.namespace=default                                                      
                                                           k8s:zgroup=kafkaTestApp                                                                      
24803      Disabled           Disabled          41358      k8s:app=empire-outpost                            f00d::a0a:0:0:60e3   10.10.0.35    ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                              
                                                           k8s:io.kubernetes.pod.namespace=default                                                      
                                                           k8s:outpostid=9999                                                                           
                                                           k8s:zgroup=kafkaTestApp                                                                      
47085      Enabled            Disabled          60020      k8s:app=kafka                                     f00d::a0a:0:0:b7ed   10.10.0.199   ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                              
                                                           k8s:io.kubernetes.pod.namespace=default                                                      
                                                           k8s:zgroup=kafkaTestApp                                                                      
53077      Disabled           Disabled          4          reserved:health                                   f00d::a0a:0:0:cf55   10.10.0.162   ready   

cmd: kubectl exec -n kube-system cilium-hvlnt -- cilium service list
ID   Frontend            Backend                   
1    10.96.44.22:9090    1 => 10.10.1.102:9090     
2    10.96.0.1:443       1 => 192.168.36.11:6443   
3    10.96.0.10:53       1 => 10.10.1.32:53        
9    10.97.166.56:2181   1 => 10.10.1.69:2181      

cmd: kubectl exec -n kube-system cilium-hvlnt -- cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                              IPv6                   IPv4          STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                                                
13949      Disabled           Disabled          55041      k8s:io.cilium.k8s.policy.serviceaccount=kube-dns         f00d::a0a:100:0:367d   10.10.1.32    ready   
                                                           k8s:io.kubernetes.pod.namespace=kube-system                                                           
                                                           k8s:k8s-app=kube-dns                                                                                  
19748      Disabled           Disabled          18096      k8s:app=zook                                             f00d::a0a:100:0:4d24   10.10.1.69    ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                       
                                                           k8s:io.kubernetes.pod.namespace=default                                                               
                                                           k8s:zgroup=kafkaTestApp                                                                               
24549      Disabled           Disabled          4          reserved:health                                          f00d::a0a:100:0:5fe5   10.10.1.224   ready   
48202      Disabled           Disabled          20446      k8s:app=empire-outpost                                   f00d::a0a:100:0:bc4a   10.10.1.90    ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                       
                                                           k8s:io.kubernetes.pod.namespace=default                                                               
                                                           k8s:outpostid=8888                                                                                    
                                                           k8s:zgroup=kafkaTestApp                                                                               
61710      Disabled           Disabled          45851      k8s:app=prometheus                                       f00d::a0a:100:0:f10e   10.10.1.102   ready   
                                                           k8s:component=core                                                                                    
                                                           k8s:io.cilium.k8s.policy.serviceaccount=prometheus-k8s                                                
                                                           k8s:io.kubernetes.pod.namespace=prometheus                                                            

===================== Existing AfterFailed =====================
```

[f0c5341e_K8sValidatedKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies (1).zip](https://github.com/cilium/cilium/files/2226747/f0c5341e_K8sValidatedKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies.1.zip)

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Upgrade impact is that the conntrack map will be recreated on startup, which may cause temporary disruption of ongoing connections.

**Comment by ianvernon** [comment]
> Upgrade impact is that the conntrack map will be recreated on startup, which may cause temporary disruption of ongoing connections. 

This is worrisome; I guess we have the issue with the L7 proxy being re-instantiated upon startup, though. 

Will this happen whenever Cilium is restarted?

**Review by ianvernon — APPROVED** [review]
Approving option changes.

File: bpf/lib/conntrack.h [hunk 1]
```diff
@@ -105,6 +120,37 @@ static inline bool __inline__ __ct_update_timeout(struct ct_entry *entry,
...
+	 * are likely to progressively set SYN, ACK, then much later perhaps
+	 * FIN and/or RST. Furthermore, unless such a traffic pattern were
+	 * constantly received, this should self-correct as the stored
+	 * tcp_flags is an OR'd set of flags and each time the above code is
+	 * executed, it pulls the latest set of accumulated flags. Therefore
+	 * even in the worst case such a conflict is likely only to cause a
+	 * small number of additional notifications, which is still likely to
+	 * be significantly less under this MONITOR_AGGREGATION mode than would
+	 * otherwise be sent if the MONITOR_AGGREGATION level is set to none
+	 * (ie, sending a notification for every packet).
```
Comment (tgraf) [inline bpf/lib/conntrack.h:152]: Thanks for these comments

**Comment by joestringer** [comment]
@ianvernon no, it only happens when the map fields change, ie during upgrade/downgrade.

TCP connections should recover shortly after upgrade as the conntrack entries will be recreated based on traffic flowing through.

File: bpf/lib/conntrack.h [hunk 2]
```diff
@@ -62,17 +62,54 @@ enum {
...
-	entry->lifetime = bpf_ktime_get_sec() + lifetime;
+	entry->lifetime = now + lifetime;
 #endif
+	if (dir == CT_INGRESS) {
+		accumulated_flags = &entry->rx_flags_seen;
+	} else {
+		accumulated_flags = &entry->tx_flags_seen;
+	}
+	seen_flags |= *accumulated_flags;
+	if (*accumulated_flags != seen_flags) {
```
Comment (tgraf) [inline bpf/lib/conntrack.h]: We might want to exclude certain flags such as ECN, URG and PSH. Not a blocker but might help prevent unnecessary notifications.

**Review by tgraf — APPROVED** [review]

**Pull request merged**